File: Tests/AINotebookCoreTests/OllamaWireTypesTests.swift
import XCTest
@testable import AINotebookCore

final class OllamaWireTypesTests: XCTestCase {
    func testChatRequestEncodes() throws {
        let req = OllamaChatRequest(
            model: "llama3.2:3b",
            messages: [
                OllamaChatMessage(role: .system, content: "You are helpful."),
                OllamaChatMessage(role: .user, content: "Hi.")
            ],
            stream: true,
            options: nil
        )
        let data = try JSONEncoder().encode(req)
        let json = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: data) as? [String: Any]
        )
        XCTAssertEqual(json["model"] as? String, "llama3.2:3b")
        XCTAssertEqual(json["stream"] as? Bool, true)
        let messages = try XCTUnwrap(json["messages"] as? [[String: Any]])
        XCTAssertEqual(messages.count, 2)
        XCTAssertEqual(messages[0]["role"] as? String, "system")
    }

    func testChatChunkDecode() throws {
        let json = """
        {"model":"llama3.2:3b","created_at":"2024-09-25T12:00:00Z","message":{"role":"assistant","content":"Hi"},"done":false}
        """.data(using: .utf8)!
        let chunk = try JSONDecoder().decode(OllamaChatChunk.self, from: json)
        XCTAssertEqual(chunk.message.content, "Hi")
        XCTAssertEqual(chunk.message.role, .assistant)
        XCTAssertFalse(chunk.done)
    }

    func testEmbedRequestEncodesArrayInput() throws {
        let req = OllamaEmbedRequest(model: "nomic-embed-text", input: ["a", "b"])
        let data = try JSONEncoder().encode(req)
        let json = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: data) as? [String: Any]
        )
        XCTAssertEqual(json["model"] as? String, "nomic-embed-text")
        XCTAssertEqual(json["input"] as? [String], ["a", "b"])
    }

    func testEmbedResponseDecodes() throws {
        let json = """
        {"embeddings":[[0.1,0.2,0.3],[0.4,0.5,0.6]]}
        """.data(using: .utf8)!
        let res = try JSONDecoder().decode(OllamaEmbedResponse.self, from: json)
        XCTAssertEqual(res.embeddings.count, 2)
        XCTAssertEqual(res.embeddings[0], [0.1, 0.2, 0.3])
    }
}

File: windows/tests/AINotebook.Core.Tests/Ollama/OllamaWireTypesTests.cs
using System.Text.Json;
using AINotebook.Core.Ollama;
using Xunit;

namespace AINotebook.Core.Tests.Ollama;

public class OllamaWireTypesTests
{
    private static readonly JsonSerializerOptions Opts = OllamaJson.Options;

    // OllamaWireTypesTests.testChatRequestEncodes
    [Fact]
    public void ChatRequestEncodesStreamTrueAndMessages()
    {
        var req = new OllamaChatRequest(
            "llama3.2:3b",
            new[]
            {
                new OllamaChatMessage(OllamaChatRole.System, "be brief"),
                new OllamaChatMessage(OllamaChatRole.User, "hi"),
            },
            Stream: true,
            Options: null);

        var json = JsonSerializer.Serialize(req, Opts);
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;

        Assert.Equal("llama3.2:3b", root.GetProperty("model").GetString());
        Assert.True(root.GetProperty("stream").GetBoolean());
        Assert.Equal(2, root.GetProperty("messages").GetArrayLength());
        Assert.Equal("system", root.GetProperty("messages")[0].GetProperty("role").GetString());
        // options is null -> omitted entirely
        Assert.False(root.TryGetProperty("options", out _));
    }

    // OllamaWireTypesTests.testChatChunkDecode (validates created_at maps)
    [Fact]
    public void ChatChunkDecodesCreatedAtAndMessage()
    {
        const string json = """
        {"model":"llama3.2:3b","created_at":"2024-09-25T12:00:00Z","message":{"role":"assistant","content":"Hi"},"done":false}
        """;
        var chunk = JsonSerializer.Deserialize<OllamaChatChunk>(json, Opts)!;
        Assert.Equal("Hi", chunk.Message.Content);
        Assert.Equal(OllamaChatRole.Assistant, chunk.Message.Role);
        Assert.False(chunk.Done);
        Assert.Equal("2024-09-25T12:00:00Z", chunk.CreatedAt);
    }

    // OllamaWireTypesTests.testEmbedRequestEncodesArrayInput
    [Fact]
    public void EmbedRequestEncodesArrayInput()
    {
        var req = new OllamaEmbedRequest("nomic-embed-text", new[] { "a", "b" });
        var json = JsonSerializer.Serialize(req, Opts);
        using var doc = JsonDocument.Parse(json);
        Assert.Equal("nomic-embed-text", doc.RootElement.GetProperty("model").GetString());
        var input = doc.RootElement.GetProperty("input");
        Assert.Equal(JsonValueKind.Array, input.ValueKind);
        Assert.Equal("a", input[0].GetString());
        Assert.Equal("b", input[1].GetString());
    }

    // OllamaWireTypesTests.testEmbedResponseDecodes
    [Fact]
    public void EmbedResponseDecodesNestedDoubleArray()
    {
        const string json = """{"embeddings":[[0.1,0.2,0.3],[0.4,0.5,0.6]]}""";
        var resp = JsonSerializer.Deserialize<OllamaEmbedResponse>(json, Opts)!;
        Assert.Equal(2, resp.Embeddings.Length);
        Assert.Equal(new[] { 0.1, 0.2, 0.3 }, resp.Embeddings[0]);
    }
}
